File: problems/33-quickselect/003-three-way-partition/python/solution.py
import random
class Solution:
    def kth_smallest(self, nums, k):
        target=k-1; left=0; right=len(nums)-1
        while left<=right:
            pivot=nums[random.randint(left,right)]
            lt=i=left; gt=right
            while i<=gt:
                if nums[i]<pivot: nums[lt],nums[i]=nums[i],nums[lt]; lt+=1; i+=1
                elif nums[i]>pivot: nums[i],nums[gt]=nums[gt],nums[i]; gt-=1
                else: i+=1
            if target<lt: right=lt-1
            elif target>gt: left=gt+1
            else: return pivot
        raise ValueError('k out of range')

File: problems/33-quickselect/003-three-way-partition/java/Solution.java
import java.util.*;
public class Solution {
    public int kthSmallest(int[] a, int k) {
        int left=0,right=a.length-1,target=k-1;
        Random random=new Random(1);
        while (left<=right) {
            int pivot=a[left+random.nextInt(right-left+1)], lt=left, i=left, gt=right;
            while(i<=gt){ if(a[i]<pivot) swap(a,lt++,i++); else if(a[i]>pivot) swap(a,i,gt--); else i++; }
            if(target<lt) right=lt-1; else if(target>gt) left=gt+1; else return pivot;
        }
        throw new IllegalArgumentException();
    }
    private void swap(int[] a,int i,int j){int t=a[i];a[i]=a[j];a[j]=t;}
}